### PR TITLE
Improve Docker configuration

### PR DESCRIPTION
**Pull request by Hritik14**
This PR creates an organized Docker environment with proper documentation inspired from scancodeio.
A `Makefile` is introduced which eases the installation greatly. Also, now `SECRET_KEY` is no longer optional. Commit messages contain further descriptive text.

With Makefile at hand, documentation of other parts also needs to be improved but that is out of realm of this PR.

### CODE REVIEW

**Comment by sbs2001** [comment]
While you're at it , maybe we could fix these misconfigs  found via https://github.com/docker/docker-bench-security .  Pasting he relevant misconfigs here.
```
[INFO] 4 - Container Images and Build File
[WARN] 4.1 - Ensure that a user for the container has been created (Automated)
[WARN]      * Running as root: vulnerablecode
[WARN]      * Running as root: vulnerablecode_db_1
[NOTE] 4.2 - Ensure that containers use only trusted base images (Manual)
[NOTE] 4.3 - Ensure that unnecessary packages are not installed in the container (Manual)
[NOTE] 4.4 - Ensure images are scanned and rebuilt to include security patches (Manual)
[WARN] 4.5 - Ensure Content trust for Docker is Enabled (Automated)
[WARN] 4.6 - Ensure that HEALTHCHECK instructions have been added to container images (Automated)
[WARN]      * No Healthcheck found: [vulnerablecode_web:latest]
[WARN]      * No Healthcheck found: a12ce5bf6f8e
[WARN]      * No Healthcheck found: 556a5c9bb265
[WARN]      * No Healthcheck found: 39bede3c7e7c
[WARN]      * No Healthcheck found: [postgres:latest]
[WARN]      * No Healthcheck found: 28a4c88cdbbf
[INFO] 4.7 - Ensure update instructions are not used alone in the Dockerfile (Manual)
[INFO]      * Update instruction found: [vulnerablecode_web:latest]
[INFO]      * Update instruction found: [postgres:latest]
[NOTE] 4.8 - Ensure setuid and setgid permissions are removed (Manual)
[INFO] 4.9 - Ensure that COPY is used instead of ADD in Dockerfiles (Manual)
[INFO]      * ADD in image history: [vulnerablecode_web:latest]
[NOTE] 4.10 - Ensure secrets are not stored in Dockerfiles (Manual)
[NOTE] 4.11 - Ensure only verified packages are are installed (Manual)

[INFO] 5 - Container Runtime
[PASS] 5.1 - Ensure that, if applicable, an AppArmor Profile is enabled (Automated)
[WARN] 5.2 - Ensure that, if applicable, SELinux security options are set (Automated)
[WARN]      * No SecurityOptions Found: vulnerablecode
[WARN]      * No SecurityOptions Found: vulnerablecode_db_1
[PASS] 5.3 - Ensure that Linux kernel capabilities are restricted within containers (Automated)
[PASS] 5.4 - Ensure that privileged containers are not used (Automated)
[PASS] 5.5 - Ensure sensitive host system directories are not mounted on containers (Automated)
[PASS] 5.6 - Ensure sshd is not run within containers (Automated)
[PASS] 5.7 - Ensure privileged ports are not mapped within containers (Automated)
[NOTE] 5.8 - Ensure that only needed ports are open on the container (Manual)
[PASS] 5.9 - Ensure that the host's network namespace is not shared (Automated)
[WARN] 5.10 - Ensure that the memory usage for containers is limited (Automated)
[WARN]       * Container running without memory restrictions: vulnerablecode
[WARN]       * Container running without memory restrictions: vulnerablecode_db_1
[WARN] 5.11 - Ensure that CPU priority is set appropriately on containers (Automated)
[WARN]       * Container running without CPU restrictions: vulnerablecode
[WARN]       * Container running without CPU restrictions: vulnerablecode_db_1
[WARN] 5.12 - Ensure that the container's root filesystem is mounted as read only (Automated)
[WARN]       * Container running with root FS mounted R/W: vulnerablecode
[WARN]       * Container running with root FS mounted R/W: vulnerablecode_db_1
[WARN] 5.13 - Ensure that incoming container traffic is bound to a specific host interface (Automated)
[WARN]       * Port being bound to wildcard IP: 0.0.0.0 in vulnerablecode
[WARN] 5.14 - Ensure that the 'on-failure' container restart policy is set to '5' (Automated)
[WARN]       * MaximumRetryCount is not set to 5: vulnerablecode
[WARN]       * MaximumRetryCount is not set to 5: vulnerablecode_db_1
[PASS] 5.15 - Ensure that the host's process namespace is not shared (Automated)
[PASS] 5.16 - Ensure that the host's IPC namespace is not shared (Automated)
[PASS] 5.17 - Ensure that host devices are not directly exposed to containers (Manual)
[INFO] 5.18 - Ensure that the default ulimit is overwritten at runtime if needed (Manual)
[INFO]       * Container no default ulimit override: vulnerablecode
[INFO]       * Container no default ulimit override: vulnerablecode_db_1
[PASS] 5.19 - Ensure mount propagation mode is not set to shared (Automated)
[PASS] 5.20 - Ensure that the host's UTS namespace is not shared (Automated)
[PASS] 5.21 - Ensurethe default seccomp profile is not Disabled (Automated)
[NOTE] 5.22 - Ensure that docker exec commands are not used with the privileged option (Automated)
[NOTE] 5.23 - Ensure that docker exec commands are not used with the user=root option (Manual)
[PASS] 5.24 - Ensure that cgroup usage is confirmed (Automated)
[WARN] 5.25 - Ensure that the container is restricted from acquiring additional privileges (Automated)
[WARN]       * Privileges not restricted: vulnerablecode
[WARN]       * Privileges not restricted: vulnerablecode_db_1
[WARN] 5.26 - Ensure that container health is checked at runtime (Automated)
[WARN]       * Health check not set: vulnerablecode
[WARN]       * Health check not set: vulnerablecode_db_1
[INFO] 5.27 - Ensure that Docker commands always make use of the latest version of their image (Manual)
[WARN] 5.28 - Ensure that the PIDs cgroup limit is used (Automated)
[WARN]       * PIDs limit not set: vulnerablecode
[WARN]       * PIDs limit not set: vulnerablecode_db_1
```

**Comment by Hritik14** [comment]
@sbs2001 These results appear to be from the old docker configs. Can you confirm if you used the configs from this branch or from the main ?

**Comment by Hritik14** [comment]
Btw this has a side effect that it fixes https://github.com/nexB/vulnerablecode/issues/462

**Comment by Hritik14** [comment]
@rolfschr can you please update the nix tests in this one

File: .travis.yml [hunk 1]
```diff
@@ -12,13 +12,12 @@ env:
   - SECRET_KEY="i1bn=oly)w*2yl-5yc&f!vvgt)p)fh3_2$r#spa!*sw36f5ov7"
   - GH_TOKEN="dummygithubtoken"
 
+# venv is already activated
 before_script:
-  - psql -c "CREATE DATABASE vulnerablecode;" -U postgres
-  - ./manage.py migrate
+  - make postgres ACTIVATE=
```
Comment (pombredanne) [inline .travis.yml:17]: What does `ACTIVATE=` mean here?

File: Dockerfile [hunk 2]
```diff
@@ -1,15 +1,11 @@
...
-# to the terminal without buffering it first
+# Force unbuffered stdout and stderr (i.e. they are flushed to terminal immediately)
 ENV PYTHONUNBUFFERED 1
-RUN mkdir /vulnerablecode
-WORKDIR /vulnerablecode
-ADD . /vulnerablecode/
-RUN pip install -r requirements.txt && \
-    DJANGO_DEV=1 python manage.py collectstatic
 
-LABEL "base_image": "pkg:docker/python@sha256%3Ae9b7e3b4e9569808066c5901b8a9ad315a9f14ae8d3949ece22ae339fff2cad0"
```
Comment (pombredanne) [inline Dockerfile:12]: Keeping labels would be a good thing

File: .travis.yml (hunk 1 above)
Comment (Hritik14) [reply .travis.yml:17]: This implies that the virtualenv is already activated i.e. the ACTIVATE command doesn't do anything (nop).
Commented in line 45.

File: Dockerfile [hunk 3]
```diff
@@ -1,15 +1,11 @@
...
-LABEL "dockerfile_url":  "https://github.com/nexB/vulnerablecode/blob/develop/Dockerfile"
-LABEL "homepage_url":  "https://github.com/nexB/vulnerablecode"
-LABEL "license": "Apache-2.0"
\ No newline at end of file
+RUN mkdir /opt/vulnerablecode && \
+		mkdir -p /var/vulnerablecode/static/
```
Comment (pombredanne) [inline Dockerfile:7]: Why mix spaces and tabs?

File: Dockerfile [hunk 4]
```diff
@@ -1,15 +1,11 @@
...
+WORKDIR /opt/vulnerablecode
+COPY . .
+RUN python -m pip install --upgrade pip && \
+		pip install -r requirements.txt
```
Comment (pombredanne) [inline Dockerfile:11]: Same: uses spaces everywhere in Dockerfiles

**Review by pombredanne — APPROVED** [review]
LGTM... I have a few questions and nits though

File: Dockerfile (hunk 3 above)
Comment (Hritik14) [reply Dockerfile:7]: ok, let's switch to spaces only.

File: Dockerfile [hunk 5]
```diff
@@ -1,15 +1,11 @@
-FROM python@sha256:e9b7e3b4e9569808066c5901b8a9ad315a9f14ae8d3949ece22ae339fff2cad0
+FROM python:3.9.6
```
[the remark this answers is not in the document]
Comment (Hritik14) [reply Dockerfile:1]: Ok, let's use  `3.8.11-slim-buster` in consistent to the CI (make CI use the same ver) Later we could decide on supported versions and create CI tests for them as well.

[the remark this answers is not in the document]
Comment (Hritik14) [reply Dockerfile]: In CI tests: Be explicit- `3.8.11-buster `
In Docker: Let's support `3.8
A test for docker-build and run

File: .travis.yml (hunk 1 above)
Comment (Hritik14) [reply .travis.yml]: As per our [discussion](https://github.com/nexB/vulnerablecode/wiki/WeeklyMeetings#meeting-on-tuesday-2021-07-27-at-1300-utc), I'll remove travisCI.

**Comment by Hritik14** [comment]
@pombredanne  nits resolved.  Merging this now

**Comment by Hritik14** [comment]
@rolfschr  ping...

**Pull request merged**